source/mir/random/discrete.d

```diff
@@ -11,6 +11,221 @@ module mir.random.discrete;
+        probs = probabilities of the individual, discrete values
+
+    Complexity: O(n), where n is the number of discrete values
+    */
+    this(const(T)[] probs)
+    {
+        debug
+        {
+            import mir.sum : sum, Summation;
+            import std.math : approxEqual;
+            assert(probs.sum!(Summation.fast).approxEqual(1.0), "Sum of the probabilities must be 1");
+        }
+        initialize(probs);
+    }
+
+    /**
+    Initialize procedure after [Vose91].
+    Each column of the probability array is either filled or used to fill a column
+    in such a order that each column has exactly two indexes - the original and
+    the alias.
+    This procedure is stable and guaranteed to find such columns as the probabilities
+    are split at the average probability p_avg and each column is picked with
+    1/n - the average probability, so larger columns that are used to fill always
+    have enough capacity.
+    Afterwards in the sampling step such a column is picked by rolling a fair dice (1/n),
+    and whether the original or alias entry in the column should be used is picked
+    by tossing a coin with the probability equal to `p_column * n`.
+    A detailed proof can be found at [Vose91].
+    */
+    private void initialize(const(T)[] probs)
+    {
+        size_t n = probs.length;
+        arr = new AltPair[probs.length];
+
+        static struct ProbsIndexed
+        {
+            T prob; // scaled probability
+            size_t index; // original column index
+        }
+
+        size_t smCounter, lgCounter;
+
+        import std.experimental.allocator.mallocator : Mallocator;
+        import std.experimental.allocator : dispose, makeArray;
+        auto alloc = Mallocator.instance;
+
+        auto small = alloc.makeArray!ProbsIndexed(n); // columns that need to be filled
+        auto large = alloc.makeArray!ProbsIndexed(n); // used to fill columns
+
+        foreach (i, p; probs)
+        {
+            auto sp = p * n;
+
+            // 1 is the average probability and depending on the ratio, we either
+            // need to add values to a block or remove values
+            if (sp < 1)
+                small[smCounter++] = ProbsIndexed(sp, i);
+            else
+                large[lgCounter++] = ProbsIndexed(sp, i);
+        }
+
+        // as long as there are elements in both stacks
+        while (smCounter > 0 && lgCounter > 0)
+        {
+            auto sm = small[--smCounter];
+            auto lg = large[--lgCounter];
+
+            // fill the smaller, discrete value with the larger
+            arr[sm.index] = AltPair(sm.prob, lg.index);
+
+            // update larger & reinsert
+            lg.prob += sm.prob - 1;
+
+            if (lg.prob < 1)
+                small[smCounter++] = lg;
+            else
+                large[lgCounter++] = lg;
+        }
+
+        // do cleanup
+        for (auto i = lgCounter; i > 0; i--)
+        {
+            arr[large[i - 1].index].prob = 1;
```

Comment:
is there a smart way to avoid `i  - 1`, e.g. should I add an extra `if lgCounter > 0` on top and do `lgCounter - 1`?